tools/src/main/java/com/nedap/archie/diff/RulesDifferentiator.java

```diff
@@ -0,0 +1,35 @@
+package com.nedap.archie.diff;
+
+import com.nedap.archie.aom.Archetype;
+import com.nedap.archie.rules.RuleStatement;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class RulesDifferentiator {
+
+    public void differentiate(Archetype result, Archetype flatParent) {
+        if (flatParent.getRules() == null
+                || flatParent.getRules().getRules() == null
+                || flatParent.getRules().getRules().size() == 0) {
+            // Parent has no rules, all rules can stay in the child archetype
+            return;
+        }
+
+        if (result.getRules().getRules().size() == flatParent.getRules().getRules().size()) {
```

Comment:
Is this a good check or is it better to just remove it and go through the for loop? It will give the same result

Reply:
The tricky part in this differentiator is that you can change rules in a flat archetype, or even remove them. This is of course not possible in the differential form of the archetype: you can only add new rules. The differentiator does not distinguish this scenario currently.

Maybe it should check if all rules from the parent should be present in the flat specialised archetype, as well as additional rules, then throw a warning or maybe even an error if a rule is missing that should be there?

If that is changed, this also means the for loop should be removed.


i also think this class needs a bit of documentation on how it works, what it does and what it does not. - in javadoc and perhaps in the Readme or the wiki of the github project.

Reply:
Good point. That also means that in general there will always be the problem that in a (flat) child archetype you can add or change rules pointing to the same element as a rule in the parent (for example a sum that is different in the child and the parent), resulting in both rules being there in the final flat archetype, which gives conflicts I assume?

So maybe such a RuleDifferentiator is not possible, or something needs to change in the Flattener before this?

Reply:
I've added some documentation.

Maybe the implications of changing and deleting rules should not be considered in the RulesDifferentiator just yet. This can be a whole project on its own because I assume most changes and deletions are not valid in specialisations.